server/src/main/java/org/opensearch/common/settings/WriteableSetting.java

```diff
@@ -0,0 +1,273 @@
+        Long,
+        Float,
+        Double,
+        String,
+        TimeValue, // long + TimeUnit
+        ByteSizeValue, // long + ByteSizeUnit
+        Version
+    }
+
+    private Setting<?> setting;
+    private SettingType type;
+
+    /**
+     * Wrap a {@link Setting}. The generic type is determined from the type of the default value.
+     *
+     * @param setting  The setting to wrap. The default value must not be null.
+     * @throws IllegalArgumentException if the setting has a null default value.
+     */
+    public WriteableSetting(Setting<?> setting) {
+        this(setting, getGenericTypeFromDefault(setting));
+    }
+
+    /**
+     * Wrap a {@link Setting} with a specified generic type.
+     *
+     * @param setting  The setting to wrap.
+     * @param type  The Generic type of the setting.
+     */
+    public WriteableSetting(Setting<?> setting, SettingType type) {
+        this.setting = setting;
+        this.type = type;
+    }
+
+    /**
+     * Wrap a {@link Setting} read from a stream.
+     *
+     * @param in Input to read the value from.
+     * @throws IOException if there is an error reading the values
+     */
+    public WriteableSetting(StreamInput in) throws IOException {
+        // Read the type
+        this.type = in.readEnum(SettingType.class);
+        // Read the key
+        String key = in.readString();
+        // Read the default value
+        Object defaultValue = readDefaultValue(in);
+        // Read a boolean specifying whether the fallback settings is null
+        WriteableSetting fallback = null;
+        boolean hasFallback = in.readBoolean();
+        if (hasFallback) {
+            fallback = new WriteableSetting(in);
+        }
+        // We are using known types so don't need the parser
+        // We are not using validator
+        // Read properties
+        EnumSet<Property> propSet = in.readEnumSet(Property.class);
+        // Put it all in a setting object
+        this.setting = createSetting(type, key, defaultValue, fallback, propSet.toArray(Property[]::new));
+    }
+
+    /**
+     * Due to type erasure, it is impossible to determine the generic type of a Setting at runtime.
+     * All settings have a non-null default, however, so the type of the default can be used to determine the setting type.
+     *
+     * @param setting The setting with a generic type.
+     * @return The corresponding {@link SettingType} for the default value.
+     */
+    private static SettingType getGenericTypeFromDefault(Setting<?> setting) {
+        String typeStr = null;
+        try {
+            // This throws NPE on null default
+            typeStr = setting.getDefault(Settings.EMPTY).getClass().getSimpleName();
+            // This throws IAE if not in enum
+            return SettingType.valueOf(typeStr);
+        } catch (NullPointerException e) {
+            throw new IllegalArgumentException("Unable to determine the generic type of this setting with a null default value.");
```

Comment:
Can we log the default values here to provide more detail?

Reply:
Missed this comment on my original review.  I actually don't think this can happen with the existing implementations due to the validators.  This is mostly a defensive catch against someone changing the code in the future, in which case I think the information given and the stack trace should provide enough detail.